open-widget-framework/test/widget_list_test.js

```diff
@@ -1,31 +1,324 @@
-      done()
+  it('makePassThroughProps constructs takes a widgetId and propagates it to the props', () => {
+    const wrap = mount(<WidgetList {...dummyProps}/>)
+    const instance = wrap.instance()
+    instance.updateWidgetList(dummyWidgetInstances)
+
+    const deleteWidgetSpy = sinon.spy(instance, 'deleteWidget')
+    const moveWidgetSpy = sinon.spy(instance, 'moveWidget')
+    const renderWidgetSpy = sinon.spy(instance, 'renderWidgetBody')
+    const renderEditWidgetFormSpy = sinon.spy(instance, 'renderEditWidgetForm')
+
+    const passThroughProps = instance.makePassThroughProps(dummyWidgetInstances[0])
+
+    passThroughProps.deleteWidget()
+    expect(deleteWidgetSpy.withArgs(dummyWidgetId).callCount).to.equal(1)
+
+    passThroughProps.moveWidget(dummyPosition)
+    expect(moveWidgetSpy.withArgs(dummyWidgetId, dummyPosition).callCount).to.equal(1)
+
+    passThroughProps.renderWidget()
+    expect(renderWidgetSpy.withArgs(dummyWidgetInstances[0]).callCount).to.equal(1)
+
+    passThroughProps.renderEditWidgetForm()
+    expect(renderEditWidgetFormSpy.withArgs(dummyWidgetId).callCount).to.equal(1)
+
+  })
+
+  it('makeFormProps constructs an appropriate set of formProps', () => {
+    const wrap = mount(<WidgetList {...dummyProps}/>)
+    const instance = wrap.instance()
+    instance.updateWidgetList(dummyWidgetInstances)
+
+    const updateWidgetListSpy = sinon.spy(instance, 'updateWidgetList')
+    const formProps = instance.makeFormProps()
+
+    expect(formProps.widgetListId).to.equal(dummyWidgetListId)
+    expect(formProps.listLength).to.equal(dummyWidgetInstances.length)
+
+    resetSpyHistory()
+    const dummyUrl = apiPath('widget', dummyWidgetId)
+    formProps.fetchData(dummyUrl)
+    expect(fetchSpy.withArgs(dummyUrl).callCount).to.equal(1)
+
+    const dummyError = 'error'
+    formProps.errorHandler(dummyError)
+    expect(dummyProps.errorHandler.withArgs(dummyError).callCount).to.equal(1)
+
+    formProps.onSubmit(dummyWidgetInstances)
+    expect(updateWidgetListSpy.withArgs(dummyWidgetInstances).callCount).to.equal(1)
+
+    const { Loader } = formProps
+    const loader = mount(<Loader />)
+    expect(loader.exists('.default-loader')).to.equal(true)
+  })
+
+  it('renderWidgetList renders a ListWrapper', () => {
+    const wrap = mount(<WidgetList {...dummyProps}/>)
+    const instance = wrap.instance()
+    instance.updateWidgetList(dummyWidgetInstances)
+
+    const listWrap = mount(instance.renderWidgetList())
+    expect(listWrap.exists('.default-list-wrapper')).to.equal(true)
+    expect(listWrap.props()).to.include(instance.makePassThroughProps())
+    expect(listWrap.props()).to.include(dummyProps.listWrapperProps)
+  })
+
+  it('renderListBody runs render widget once for each widget in widget instances', () => {
+    const wrap = mount(<WidgetList {...dummyProps}/>)
+    const instance = wrap.instance()
+    instance.updateWidgetList(dummyWidgetInstances)
```

Comment:
Can you add a test where `updateWidgetList` is not called before running `renderListBody`? When testing via the sample app I noticed that the `widgetInstances` state is sometimes null. Is it possible to change this to never be null, or otherwise to make sure we're checking null status in each function?

Reply:
The reason we're using `null` is to be able to differentiate between waiting for the async data to be returned and being given an empty list. We could use `undefined` if that's any better. I worry that if we set it to an empty array, it will muck with the the wrapper and may not refresh when the async function does return (for example, we would have to run makePassThroughProps again when the data returns or the props passed to ListWrapper are inaccurate).

Reply:
@noisecapella This is the last remaining issue on the PR

Reply:
I'm fine with using `null` in that way, but there should be tests for `updateWidgetList` where `widgetInstances` is null so we can verify that the case is handled correctly.